Rework explainer to use the Login Status API (#52)
fedidcg/LightweightFedCM: README.md

## Goals and Non-Goals

The following use cases are all motivating to this work and it is our goal to provide an easy-to-integrate solution for them that can be integrated into the Credential Manager as a unified browser-mediated login mechanism.

This feature is intended to:
* Allow IdPs to request Storage Access with a UI that gives users more context for their choices.
* Provide the benefits of the FedCM UX to IdPs whose system architecture or engineering budget does not permit adoption of full FedCM.
* Improve performance and privacy properties of FedCM by moving credentialed calls to IdPs to a later point in the sign-in user journey.
* Act as a supplement to FedCM, not an entirely parallel API surface; ideally, an understanding of FedCM should confer some understanding of Lightweight FedCM, and vice versa.
This feature explicitly must /not/:
* Make invisible/silent timing attacks possible for a colluding IdP and RP.
* Create incentives for companies engaged in tracking to present themselves as a fake IdP.
This feature is /not/ intended to:

> * Entirely eliminate the need for some serving changes on the IdP; a limited number of new endpoints are acceptable in order to support some functionality.

Author: I was explicitly trying to rule out "no infrastructure changes on the IdP" as a requirement in our problem statement here, but it seems that's still at issue :)

> * Entirely eliminate the need for some serving changes on the IdP; a limited number of new endpoints are acceptable in order to support some functionality.

Reviewer: Whoops! My brain skipped the /not/ above! Never mind!

FedCM, by default, makes a distinction between sign-up and sign-in via a property called approved_clients in the user profile information: if the clientId passed in the JS call is not a member of approved_clients, it means that this client was never approved previously by the user.

This could also be accomplished client-side only by passing a list of approved_clients into the login status API call:

```js
navigator.login.setStatus("logged-in", {
	accounts: [{
         // ... other fields
         approved_clients: ["https://rp.example"],
         // ... other fields
      }]
});
```

The prompt also takes the RPs Privacy Policy and Terms of Service, which would normally be passed via the client_metadata_endpoint, but in the absence of one, could be passed in the JS call by the RP. This has also been [suggested in discussions around the IdP Registration API](https://github.com/w3c-fedid/idp-registration/issues/8).

```js
const {profile: {id}} = await navigator.credentials.get({
  identity: {
        providers: [{url: "https://example.com"}],
        rp: {
          privacyPolicy: "https://rp.example/pp.html",
          termsOfService: "https://rp.example/tos.html",

>         }

Reviewer: @ekovac This is what I meant

Author: ack. I've added some details about this to the PR.

#### `branding` ✔️

Supplying this in `n.l.setStatus` should be fine, with the caveat that the icon URLs provided must be retrieved when `setStatus` is called, *not* when the RP calls `n.c.get`.

#### `login_url` ❗

The primary caveat is that if a `login_url` is supplied in `setStatus`, it may include identifying information for the user before the user has allowed linking between the RP and IdP. This can be partially alleviated by not supporting an RP-supplied `loginHint` or `domainHint` that would allow the IdP to create that linkage too early in the user journey for user permission to be gathered. The user agent may also choose to require additional confirmation before navigation to the loginUrl in this case.

#### Distinguishing between `sign-up` and `sign-in` without an `accounts_endpoint`

FedCM, by default, makes a distinction between sign-up and sign-in via a property called approved_clients in the user profile information that is received in the [fetch the accounts] step: [https://w3c-fedid.github.io/FedCM/#fetch-the-accounts](): if the clientId passed in the JS call is not a member of approved_clients, it means that this client was never approved previously by the user.

> This could also be accomplished client-side by passing a list of approved_clients into the login status API call:

Reviewer 2: ```suggestion
This could also be accomplished client-side by passing a list of `approved_clients` in the login status API call:
```

```js
navigator.login.setStatus("logged-in", {
	accounts: [{
         // ... other fields
         approved_clients: ["https://rp.example"],
         // ... other fields
      }]
});
```

### Handling Selective Disclosure with the Field Selector

> An RP can also select which attributes of the user profile they would like the IdP or user agent to disclose. Currently, the following values are allowed: name, email, picture, phoneNumber and empty list. This aligns with [the current custom requests proposal](https://github.com/w3c-fedid/custom-requests/issues/4).
> 
> This would also work client-side only, with the returned credential object only containing the specific attributes that were requested if they were stored in the setStatus call.

Reviewer 2: ```suggestion
An RP can also select which attributes of the user profile they would like the IdP or user agent to disclose. Currently, the following values are allowed: `name`, `email`, `picture`, `phoneNumber`, and empty list. This aligns with [the current custom requests proposal](https://github.com/w3c-fedid/custom-requests/issues/4).

This would also work client-side only, with the returned credential object only containing the specific attributes that were requested if they were stored in the `setStatus` call.
```

```js
const {profile: {email, name}} = await navigator.credentials.get({
  identity: {
        providers: [{
          url: "https://example.com",
          fields: ["email", "name"] // doesn't request the profile picture
        }],
      }
});
```

### Handling Multiple IdPs

> This would extend well to multiple IdPs: if the user was "logged-in" in multiple IdPs, an RP could call:

Reviewer 2: ```suggestion
This would extend well if the user was "logged-in" in multiple IdPs, when an RP call could be as follows:
```

const token = (await fetch("https://example.com/login?rp=rp.example&id=" + id)).text();
// Do whatever you would usually do with your token here.
```

This approach works well with the Passive mode: if the user is not already signed in to the IdP, the browser doesn’t need to display anything. In Active Mode, the RP may wish to provide some guidance or trigger some action to get the user to refresh their IdP session, like navigating to the login page if it is known.

> [!NOTE]
> We refer to a `url` field on the RP-side provider configurations instead of `configURL`; this is to align with future work deprecating the 
> `configURL` parameter in favor of `url` in FedCM.

### Active Mode: Handling the Signed-Out Case

For Active mode, if you want to be able to gracefully handle the signed-out case, IdPs have an incremental path: allow users to login using the “url” passed by the RP or (optionally) the IdP providing a custom one.

So, if an RP called:

```
let cred = await navigator.credentials.get({
	identity: {
         mode: "active",
         providers: [{url: "https://example.com"}]
       },
});
```

The browser would check the sign-in status list, 

> first by trying to resolve “https://example.com/.well-known/web-identity”. If the well-known file doesn’t exist (i.e., the request returns a status 404), the browser would use `“login_url”` as the value of `“https://” + eltd+1(“url”)` (i.e., in this example, `“https://example.com”`). If an IdP requires a custom `login_url`, they can have a very simple `.well-known/identity` file, such as the following:

Reviewer: A further alternative could be to allow navigation to the full URL if the user accepts an IDP-dialog if there is some indication the url is actually a participating IDP (like prior registration or a origin-level well-known)

> first by trying to resolve “https://example.com/.well-known/web-identity”. If the well-known file doesn’t exist (i.e., the request returns a status 404), the browser would use `“login_url”` as the value of `“https://” + eltd+1(“url”)` (i.e., in this example, `“https://example.com”`). If an IdP requires a custom `login_url`, they can have a very simple `.well-known/identity` file, such as the following:

Author: Just put "when do we show an IdP picker/dialog and when is it UA-defined?" in the open questions.

### `id_assertion_endpoint` ✔️

There are no special considerations for this, since this is only invoked after the user has confirmed that they want to link their identity between the RP and IdP.

### `disconnect_endpoint` ✔️

Supplying this in `n.l.setStatus` should be fine, since this will only be invoked after a user has linked their account in the first place.

### `branding` ✔️

Supplying this in `n.l.setStatus` should be fine, with the caveat that the icon URLs provided must be retrieved when `setStatus` is called, *not* when the RP calls `n.c.get`.

### `login_url` ❗

The primary caveat is that if a `login_url` is supplied in `setStatus`, it could include identifying information for the user before the user has allowed linking between the RP and IdP. This can be partially alleviated by not supporting an RP-supplied `loginHint` or `domainHint` that would allow the IdP to create that linkage too early in the user journey for user permission to be gathered. The user agent can also choose to require additional confirmation before navigating to the loginUrl.

## Interaction with other FedCM Features

### "Login-Status" Headers

> An IdP sending `Login-Status: logged-in` refreshes the expiration timer on the stored profile information.

Reviewer 3: Could you add a little more detail to this? Why do we think that this is the right choice, and how does it refresh it? Just reset the counter?

Author: Added some detail about this, as well as more detail in the open questions about how it should interact with the header.

- Achim Schlosser
- Phil Smart
- Martin Thompson
- Christian Biesinger

> - Erica Kovac

Reviewer: Yeah, that sounds good!
